From 895fa8a9cc0764eb6d44c43514f614cdc474d1dc Mon Sep 17 00:00:00 2001
From: Elmar Pruesse <elmar.pruesse@ucdenver.edu>
Date: Tue, 2 Apr 2019 13:15:22 -0600
Subject: [PATCH 1/3] Build dylib on MacOS

- Using "ar cr" yields architecture errors with default toolchain, using
  "libtool -static -o" instead seems to work.
- On MacOS, libs end in "dylib"
- On MacOS, SOVERSION comes between basename and extension.
---
 Makefile | 25 +++++++++++++++++--------
 1 file changed, 17 insertions(+), 8 deletions(-)

diff --git a/Makefile b/Makefile
index 3013b6b..918f22f 100644
--- a/Makefile
+++ b/Makefile
@@ -41,14 +41,17 @@ BINDIR ?= $(PREFIX)/bin
 INCDIR ?= $(PREFIX)/include
 LIBDIR ?= $(PREFIX)/lib
 
-SOVERSION          := 0
+SOVERSION          := .0
 STATIC_LIB_SUFFIX  := .a
-SHARED_LIB_SUFFIX  := .so.$(SOVERSION)
+SHARED_LIB_SUFFIX  := .so
+SHARED_LIB          = libdeflate$(SHARED_LIB_SUFFIX)$(SOVERSION)
 SHARED_LIB_CFLAGS  := -fPIC
-SHARED_LIB_LDFLAGS := -Wl,-soname=libdeflate$(SHARED_LIB_SUFFIX)
+SHARED_LIB_LDFLAGS := -Wl,-soname=$(SHARED_LIB)
 PROG_SUFFIX        :=
 PROG_CFLAGS        :=
 HARD_LINKS         := 1
+ARFLAGS            := cr
+ARCHS              := *
 
 # Compiling for Windows with MinGW?
 ifneq ($(findstring -mingw,$(shell $(CC) -dumpmachine 2>/dev/null)),)
@@ -70,6 +73,13 @@ ifneq ($(findstring -mingw,$(shell $(CC) -dumpmachine 2>/dev/null)),)
         AR := $(shell echo $(CC) | \
                 sed -E 's/g?cc(-?[0-9]+(\.[0-9]+)*)?(\.exe)?$$/ar\3/')
     endif
+else ifeq ($(shell uname),Darwin)
+   SHARED_LIB_SUFFIX  := .dylib
+   SHARED_LIB          = libdeflate$(SOVERSION)$(SHARED_LIB_SUFFIX)
+   SHARED_LIB_LDFLAGS := "-install_name $(SHARED_LIB)"
+   AR                 := libtool
+   ARFLAGS            := -static -o
+   ARCHS              := x86
 endif
 
 ##############################################################################
@@ -95,14 +105,13 @@ DEFAULT_TARGETS :=
 #### Library
 
 STATIC_LIB := libdeflate$(STATIC_LIB_SUFFIX)
-SHARED_LIB := libdeflate$(SHARED_LIB_SUFFIX)
 
 LIB_CFLAGS += $(CFLAGS) -fvisibility=hidden -D_ANSI_SOURCE
 
 LIB_HEADERS := $(wildcard lib/*.h) $(wildcard lib/*/*.h)
 
 LIB_SRC := lib/aligned_malloc.c lib/deflate_decompress.c \
-	   $(wildcard lib/*/cpu_features.c)
+	   $(wildcard lib/$(ARCHS)/cpu_features.c)
 
 DECOMPRESSION_ONLY :=
 ifndef DECOMPRESSION_ONLY
@@ -139,7 +148,7 @@ $(SHARED_LIB_OBJ): %.shlib.o: %.c $(LIB_HEADERS) $(COMMON_HEADERS) .lib-cflags
 
 # Create static library
 $(STATIC_LIB):$(STATIC_LIB_OBJ)
-	$(QUIET_AR) $(AR) cr $@ $+
+	$(QUIET_AR) $(AR) $(ARFLAGS) $@ $+
 
 DEFAULT_TARGETS += $(STATIC_LIB)
 
@@ -152,9 +161,9 @@ DEFAULT_TARGETS += $(SHARED_LIB)
 
 ifdef SOVERSION
 # Create the symlink libdeflate.so => libdeflate.so.$SOVERSION
-libdeflate.so:$(SHARED_LIB)
+libdeflate$(SHARED_LIB_SUFFIX):$(SHARED_LIB)
 	$(QUIET_LN) ln -sf $+ $@
-DEFAULT_TARGETS += libdeflate.so
+DEFAULT_TARGETS += libdeflate$(SHARED_LIB_SUFFIX)
 endif
 
 # Rebuild if CC, LIB_CFLAGS, or CPPFLAGS changed
-- 
2.14.0


From ac465c6c4247ca15f54afb984a4689d473504881 Mon Sep 17 00:00:00 2001
From: Elmar Pruesse <elmar.pruesse@ucdenver.edu>
Date: Tue, 2 Apr 2019 13:51:38 -0600
Subject: [PATCH 2/3] Use ar --target mach-o-x86-64 instead of libtool

---
 Makefile | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index 918f22f..bc0a4d5 100644
--- a/Makefile
+++ b/Makefile
@@ -77,8 +77,7 @@ else ifeq ($(shell uname),Darwin)
    SHARED_LIB_SUFFIX  := .dylib
    SHARED_LIB          = libdeflate$(SOVERSION)$(SHARED_LIB_SUFFIX)
    SHARED_LIB_LDFLAGS := "-install_name $(SHARED_LIB)"
-   AR                 := libtool
-   ARFLAGS            := -static -o
+   ARFLAGS            += --target mach-o-x86-64
    ARCHS              := x86
 endif
 
-- 
2.14.0


From 781d9ea406771b278b078eb3102ed60f7f400dda Mon Sep 17 00:00:00 2001
From: Elmar Pruesse <elmar.pruesse@ucdenver.edu>
Date: Tue, 2 Apr 2019 14:21:59 -0600
Subject: [PATCH 3/3] Remove AR and ARCH overrides - just wrong ar

---
 Makefile | 2 --
 1 file changed, 2 deletions(-)

diff --git a/Makefile b/Makefile
index bc0a4d5..572ecd0 100644
--- a/Makefile
+++ b/Makefile
@@ -77,8 +77,6 @@ else ifeq ($(shell uname),Darwin)
    SHARED_LIB_SUFFIX  := .dylib
    SHARED_LIB          = libdeflate$(SOVERSION)$(SHARED_LIB_SUFFIX)
    SHARED_LIB_LDFLAGS := "-install_name $(SHARED_LIB)"
-   ARFLAGS            += --target mach-o-x86-64
-   ARCHS              := x86
 endif
 
 ##############################################################################
-- 
2.14.0

